File: apps/android/Components/NostrContent/NostrContentGrouping.kt
// Requires: nothing beyond the Kotlin stdlib. Kotlin 1.9+.
//
// Block-level grouping for a `ContentTreeWire`. `NostrContentView` consumes
// these in order; the grouping logic lives in its own file so the renderer
// composable stays focused on layout.

package nmp.content

/** Display level for inline-text groups; lets the renderer pick paragraph vs
 *  heading typography without re-walking the arena. */
public sealed class NostrContentInlineLevel {
    public object Paragraph : NostrContentInlineLevel()
    public data class Heading(val level: UByte) : NostrContentInlineLevel()
}

/** Block-level group produced by walking the roots of a [ContentTreeWire]. */
public sealed class NostrContentGroup {
    /** Inline runs that should be concatenated. The [children] are node
     *  indices; `UInt.MAX_VALUE` is a sentinel trailing newline used to
     *  separate adjacent paragraphs / headings. */
    public data class Inline(
        val level: NostrContentInlineLevel,
        val children: List<UInt>,
    ) : NostrContentGroup()

    public data class MediaGroup(val urls: List<String>, val kind: MediaKind) : NostrContentGroup()
    public data class EventRefGroup(val uri: WireNostrUri) : NostrContentGroup()
    public data class CodeBlockGroup(val info: String?, val body: String) : NostrContentGroup()
    public data class BlockQuoteGroup(val children: List<UInt>) : NostrContentGroup()
    public data class ListGroup(
        val orderedStart: ULong?,
        val items: List<List<UInt>>,
    ) : NostrContentGroup()
    public object RuleGroup : NostrContentGroup()
    public data class ImageGroup(
        val alt: String,
        val title: String?,
        val src: String?,
    ) : NostrContentGroup()
    public data class PlaceholderGroup(val reason: PlaceholderReason) : NostrContentGroup()
}

/** Sentinel index used to inject a trailing newline between contiguous inline
 *  groups (paragraph / heading). `UInt.MAX_VALUE` cannot reference a real
 *  arena slot — the bounds check in [ContentTreeWire.nodeAt] rejects it — so
 *  renderers can safely treat it as a newline marker. */
public val NOSTR_CONTENT_NEWLINE_SENTINEL: UInt = UInt.MAX_VALUE

/**
 * Flatten the document into a sequence of block-level groups. Mirrors the
 * `nostrContentGroups` function in the SwiftUI registry. Inline event refs
 * appearing inside paragraphs are pulled out into their own `EventRefGroup`
 * so the quote card renders block-level.
 */
public fun nostrContentGroups(tree: ContentTreeWire): List<NostrContentGroup> {
    val groups = mutableListOf<NostrContentGroup>()
    var pendingChildren = mutableListOf<UInt>()
    var pendingLevel: NostrContentInlineLevel = NostrContentInlineLevel.Paragraph

    fun flush() {
        if (pendingChildren.isNotEmpty()) {
            groups.add(NostrContentGroup.Inline(pendingLevel, pendingChildren.toList()))
            pendingChildren = mutableListOf()
            pendingLevel = NostrContentInlineLevel.Paragraph
        }
    }

    fun appendInline(level: NostrContentInlineLevel, children: List<UInt>) {
        if (pendingChildren.isNotEmpty() && pendingLevel != level) {
            flush()
        }
        pendingLevel = level

        val startCount = pendingChildren.size
        for (child in children) {
            val childNode = tree.nodeAt(child) ?: continue
            if (childNode is WireNode.EventRef) {
                flush()
                groups.add(NostrContentGroup.EventRefGroup(childNode.uri))
            } else {
                pendingChildren.add(child)
            }
        }
        if (pendingChildren.size > startCount) {
            pendingChildren.add(NOSTR_CONTENT_NEWLINE_SENTINEL)
        }
    }

    for (root in tree.roots) {
        val node = tree.nodeAt(root) ?: continue
        when (node) {
            is WireNode.Paragraph -> appendInline(NostrContentInlineLevel.Paragraph, node.children)
            is WireNode.Heading -> {
                flush()
                appendInline(NostrContentInlineLevel.Heading(node.level), node.children)
                flush()
            }
            is WireNode.Media -> {
                flush()
                groups.add(NostrContentGroup.MediaGroup(node.urls, node.mediaKind))
            }
            is WireNode.EventRef -> {
                flush()
                groups.add(NostrContentGroup.EventRefGroup(node.uri))
            }
            is WireNode.CodeBlock -> {
                flush()
                groups.add(NostrContentGroup.CodeBlockGroup(node.info, node.body))
            }
            is WireNode.BlockQuote -> {
                flush()
                groups.add(NostrContentGroup.BlockQuoteGroup(node.children))
            }
            is WireNode.ListNode -> {
                flush()
                groups.add(NostrContentGroup.ListGroup(node.orderedStart, node.items))
            }
            WireNode.Rule -> {
                flush()
                groups.add(NostrContentGroup.RuleGroup)
            }
            is WireNode.Image -> {
                flush()
                groups.add(NostrContentGroup.ImageGroup(node.alt, node.title, node.src))
            }
            is WireNode.Placeholder -> {
                flush()
                groups.add(NostrContentGroup.PlaceholderGroup(node.reason))
            }
            // Bare inline node at the root (plain text without paragraph wrap).
            // Treat it as a one-child paragraph.
            is WireNode.Text,
            is WireNode.Mention,
            is WireNode.Hashtag,
            is WireNode.Url,
            is WireNode.Emoji,
            is WireNode.Invoice,
            is WireNode.Emphasis,
            is WireNode.Strong,
            is WireNode.InlineCode,
            is WireNode.Link,
            WireNode.SoftBreak,
            WireNode.HardBreak -> appendInline(NostrContentInlineLevel.Paragraph, listOf(root))
        }
    }
    flush()
    return groups
}

File: apps/ios/Components/NostrContent/NostrContentGrouping.swift
import SwiftUI

/// Block-level group produced by walking the roots of a `ContentTreeWire`.
/// `NostrContentView` consumes these in order; the grouping logic exists in a
/// separate file so the renderer view itself stays focused on layout.
public enum NostrContentGroup: Equatable {
    /// Inline runs that should be concatenated with `Text + Text`. The
    /// `[UInt32]` values are node indices; `.init(UInt32.max)` is a sentinel
    /// trailing newline used to separate adjacent paragraphs/headings.
    case inline(level: NostrContentInlineLevel, children: [UInt32])
    case media(urls: [String], kind: NostrMediaKind)
    case eventRef(NostrWireUri)
    case codeBlock(info: String?, body: String)
    case blockQuote(children: [UInt32])
    case list(orderedStart: UInt64?, items: [[UInt32]])
    case rule
    case image(alt: String, title: String?, src: String?)
    case placeholder(reason: NostrWirePlaceholderReason)
}

/// Display level for inline-text groups; lets the renderer pick paragraph vs
/// heading typography without re-walking the arena.
public enum NostrContentInlineLevel: Equatable {
    case paragraph
    case heading(level: UInt8)
}

/// Sentinel index used to inject a trailing newline between contiguous inline
/// groups (paragraph/heading). Index `UInt32.max` cannot reference a real
/// arena slot — the bounds check in `node(at:)` rejects it — so renderers can
/// safely treat it as a newline marker.
public let nostrContentNewlineSentinel: UInt32 = .max

/// Flatten the document into the block-level sequence consumed by
/// `NostrContentView`.
public func nostrContentGroups(_ tree: ContentTreeWire) -> [NostrContentGroup] {
    var groups: [NostrContentGroup] = []
    var pendingChildren: [UInt32] = []
    var pendingLevel: NostrContentInlineLevel = .paragraph

    func flush() {
        if !pendingChildren.isEmpty {
            groups.append(.inline(level: pendingLevel, children: pendingChildren))
            pendingChildren = []
            pendingLevel = .paragraph
        }
    }

    func appendInline(level: NostrContentInlineLevel, children: [UInt32]) {
        // Different inline-block levels can't share a paragraph; flush first.
        if !pendingChildren.isEmpty, pendingLevel != level {
            flush()
        }
        pendingLevel = level

        let startCount = pendingChildren.count
        for child in children {
            guard let childNode = tree.node(at: child) else { continue }
            if case .eventRef(let uri) = childNode {
                flush()
                groups.append(.eventRef(uri))
            } else {
                pendingChildren.append(child)
            }
        }
        if pendingChildren.count > startCount {
            pendingChildren.append(nostrContentNewlineSentinel)
        }
    }

    for root in tree.roots {
        guard let node = tree.node(at: root) else { continue }
        switch node {
        case .paragraph(let children):
            appendInline(level: .paragraph, children: children)
        case .heading(let level, let children):
            flush()
            appendInline(level: .heading(level: level), children: children)
            flush()
        case .media(let urls, let kind):
            flush()
            groups.append(.media(urls: urls, kind: kind))
        case .eventRef(let uri):
            flush()
            groups.append(.eventRef(uri))
        case .codeBlock(let info, let body):
            flush()
            groups.append(.codeBlock(info: info, body: body))
        case .blockQuote(let children):
            flush()
            groups.append(.blockQuote(children: children))
        case .list(let orderedStart, let items):
            flush()
            groups.append(.list(orderedStart: orderedStart, items: items))
        case .rule:
            flush()
            groups.append(.rule)
        case .image(let alt, let title, let src):
            flush()
            groups.append(.image(alt: alt, title: title, src: src))
        case .placeholder(let reason):
            flush()
            groups.append(.placeholder(reason: reason))
        default:
            // Bare inline node sitting at the root (e.g. plain text with no
            // surrounding paragraph). Treat it as a one-child paragraph.
            appendInline(level: .paragraph, children: [root])
        }
    }
    flush()
    return groups
}
